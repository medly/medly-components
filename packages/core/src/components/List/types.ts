import { HTMLProps } from '@medly-components/utils';

export interface ListProps extends HTMLProps<HTMLUListElement> {
    /** List items direction */
    variant?: 'horizontal' | 'vertical';
}
