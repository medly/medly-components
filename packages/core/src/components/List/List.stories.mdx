import { List } from './List';
import { defaultTheme } from '@medly-components/theme';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './List.stories';

<Meta title="Core" component={List} />

# List

The `List` component allows you to display a list of elemsts like a paragraph, links, headings, etc.


<Preview withToolbar>
    <Story name="List">{stories.Basic()}</Story>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={List} />


