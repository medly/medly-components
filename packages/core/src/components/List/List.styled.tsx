/* stylelint-disable  no-descending-specificity */
import { defaultTheme, SpacingTheme } from '@medly-components/theme';
import { css, styled } from '@medly-components/utils';
import { ListProps } from './types';

const horizontal = (spacings: SpacingTheme) => css`
    li {
        margin: 0 ${spacings.S1};
        &:last-child {
            margin-right: ${spacings.S2};
        }
        &:first-child {
            margin-left: ${spacings.S2};
        }
    }
`;

const vertical = (spacings: SpacingTheme) => css`
    li {
        margin: ${spacings.S1} 0;
        &:last-child {
            margin-bottom: ${spacings.S2};
        }
        &:first-child {
            margin-top: ${spacings.S2};
        }
    }
`;

export const ListStyled = styled('ul').attrs(({ theme: { spacing: spacings } }) => ({ spacings }))<ListProps>`
    padding: 0;
    margin: 0;
    display: flex;
    list-style: none;
    flex-direction: ${({ variant }) => (variant === 'horizontal' ? 'row' : 'column')};
    ${({ spacings, variant }) => variant === 'horizontal' && horizontal(spacings)};
    ${({ spacings, variant }) => variant === 'vertical' && vertical(spacings)};
`;

ListStyled.defaultProps = {
    theme: defaultTheme
};
