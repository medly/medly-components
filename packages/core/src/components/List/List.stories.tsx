import { select } from '@storybook/addon-knobs';
import React from 'react';
import Text from '../Text';
import { List } from './List';
import { ListProps } from './types';

const variants: Required<ListProps>['variant'][] = ['horizontal', 'vertical'];

export const Basic = () => (
    <List variant={select('Variant', variants, 'vertical')}>
        <Text> Item 1</Text>
        <Text> Item 2</Text>
        <Text> Item 3</Text>
    </List>
);
