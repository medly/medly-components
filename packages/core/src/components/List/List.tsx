import { WithStyle } from '@medly-components/utils';
import React, { FC } from 'react';
import { ListStyled } from './List.styled';
import { ListProps } from './types';

const Component: FC<ListProps> = React.memo(
    React.forwardRef((props, ref) => (
        <ListStyled ref={ref} {...props}>
            {React.Children.map(props.children, (c, i) => {
                return <li key={i}>{c}</li>;
            })}
        </ListStyled>
    ))
);

Component.displayName = 'List';
Component.defaultProps = {
    variant: 'vertical'
};
export const List: FC<ListProps> & WithStyle = Object.assign(Component, { Style: ListStyled });
