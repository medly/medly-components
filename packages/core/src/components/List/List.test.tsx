import { render } from '@test-utils';
import { List } from './List';

describe('List component', () => {
    it('should render horizontal list', () => {
        const { container } = render(
            <List variant="horizontal">
                <p>item 1</p>
                <p>item 2</p>
            </List>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render vertical list', () => {
        const { container } = render(
            <List variant="vertical">
                <p>item 1</p>
                <p>item 2</p>
            </List>
        );
        expect(container).toMatchSnapshot();
    });
});
