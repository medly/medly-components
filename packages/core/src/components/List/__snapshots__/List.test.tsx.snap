// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render horizontal list 1`] = `
.c0 {
  padding: 0;
  margin: 0;
  display: flex;
  list-style: none;
  flex-direction: row;
}

.c0 li {
  margin: 0 0.4rem;
}

.c0 li:last-child {
  margin-right: 0.8rem;
}

.c0 li:first-child {
  margin-left: 0.8rem;
}

<div>
  <ul
    class="c0"
    variant="horizontal"
  >
    <li>
      <p>
        item 1
      </p>
    </li>
    <li>
      <p>
        item 2
      </p>
    </li>
  </ul>
</div>
`;

exports[`List component should render vertical list 1`] = `
.c0 {
  padding: 0;
  margin: 0;
  display: flex;
  list-style: none;
  flex-direction: column;
}

.c0 li {
  margin: 0.4rem 0;
}

.c0 li:last-child {
  margin-bottom: 0.8rem;
}

.c0 li:first-child {
  margin-top: 0.8rem;
}

<div>
  <ul
    class="c0"
    variant="vertical"
  >
    <li>
      <p>
        item 1
      </p>
    </li>
    <li>
      <p>
        item 2
      </p>
    </li>
  </ul>
</div>
`;
